### DISCUSSION TITLE
Search in offline mode

### DISCUSSION THREAD

**Discussion by niskrev** [quarto-dev/quarto-cli#8035] [Q&A] [answered]
### Description

Is it possible to enable search in rendered documents (website, blog, book) in offline mode? The search box appears but shows no results.

**Comment by mcanouil** [comment]
It should work.
I've just tried locally and offline with my own website.
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/477c0e5b-2ba7-4590-a9aa-7fb0bc3c8922">

**Reply by niskrev** [reply]
It doesn't work when I open _site/index.html in the browser. I am interested in using this for offline documentation, similar to [MkDocs](https://squidfunk.github.io/mkdocs-material/setup/building-for-offline-usage/)

![image](https://github.com/quarto-dev/quarto-cli/assets/20889079/233a0ee7-8780-4e15-b4e2-001bc61ab47b)

**Reply by mcanouil** [reply]
I don't know what you are doing, but it works perfectly offline.
1. I turned off network connection.
2. `quarto create project website demo`

See the results

https://github.com/quarto-dev/quarto-cli/assets/8896044/2de708e7-22ff-4ba9-81a1-f83ae06ab5f0


Note that JavaScript cannot run from `file` protocol.
<img width="1512" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/4d691190-c123-4660-b2ee-1bb976ef9c08">

**Reply by niskrev** [reply]
As I said, the search doesn't work when I open _site/index.html in the browser. This is what I mean by offline mode - without starting a local preview server. Offline search is available in MkDocs and Jupyter Book. My question is whether it is available with Quarto as well. Thanks!

**Reply by cscheid** [reply, marked as answer]
FYI, A large number of modern javascript features simply don't work when you use the `file://` protocol. That's by design, because of security concerns. As a result, a number of quarto features don't work either. Search and OJS code cells are some of the examples.

MkDocs and JupyterBook probably are using a different searching library that doesn't hit the javascript security limitations. I would encourage you to use the preview server (or a local server like the one that is bundled with Python, via `python -m http.server`).

**Reply by niskrev** [reply]
So the answer is no. Thanks for clarifying

**Reply by cscheid** [reply]
More accurately, _you're not using offline mode_. You can still be offline, disconnected from the internet, and have local search working. You do need an HTTP server (local or not).

**Reply by LJFRIESE** [reply]
For future scavengers, there is a workaround:

Step 1: After building the site, comment out the following in `site_libs/quarto-search/quarto-search.js`

```
if (window.location.protocol === "file:" && !shownWarning) {
      window.alert(
        "Search requires JavaScript features disabled when running in file://... URLs. In order to use search, please run this document in a web server."
      );
      shownWarning = true;
      return;
    }
```

Step 2: Create a .bat file in the site root directory with the following

```
REM Kill all existing instance of chrome 
taskkill /F /IM chrome.exe /T
REM directory path where chrome.exe is located
set chromeLocation="C:\Program Files (x86)\Google\Chrome\Application"
cd %chromeLocation%
cd c:
start chrome.exe "{path to index}\index.html" --allow-file-access-from-files
```
**_NB_**: See [this SO discussion](https://stackoverflow.com/questions/18586921/how-to-launch-html-using-chrome-at-allow-file-access-from-files-mode) for an explanation of what the `--allow-file-access-from-files` tag does and why you shouldn't use the window opened by this script for web browsing.
